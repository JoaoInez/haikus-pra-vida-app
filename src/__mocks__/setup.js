jest.mock("react-native/Libraries/Animated/src/NativeAnimatedHelper");
jest.setTimeout(30000);
