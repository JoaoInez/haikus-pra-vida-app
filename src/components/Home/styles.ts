import { StyleSheet } from "react-native";

const fade = StyleSheet.create({
  default: {
    fontFamily: "UnifrakturMaguntia",
    fontSize: 40,
    textAlign: "center",
    position: "absolute",
  }
})

export const defaultStyles = StyleSheet.create({
  titleContainer: {
    position: "relative"
  },
  fade1: {
    ...fade.default,
    color: "#FF006E",
    zIndex: -1,
    top: 2,
    left: 2,
  },
  fade2: {
    ...fade.default,
    color: "#2A9D8F",
    zIndex: -2,
    top: 4,
    left: 4,
  },
  fade3: {
    ...fade.default,
    color: "#FFBE0B",
    zIndex: -3,
    top: 6,
    left: 6,
  },
  fade4: {
    ...fade.default,
    color: "#3A86FF",
    zIndex: -4,
    top: 8,
    left: 8,
  },
  fade5: {
    ...fade.default,
    color: "#8338EC",
    zIndex: -5,
    top: 10,
    left: 10,
  },
  textContainer: {
    alignItems: "flex-start",
    position: "relative",
  },
  shareContainer: {
    marginTop: 20,
    flexDirection: "row",
    alignSelf: "center",
  },
  loadingContainer: {
    position: "absolute",
    top: 25,
    right: 25,
  },
});

export const lightStyles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "space-between",
    paddingVertical: 80,
  },
  title: {
    fontFamily: "UnifrakturMaguntia",
    fontSize: 40,
    textAlign: "center",
  },
  text: {
    fontSize: 18,
  },
  shareBtn: {
    padding: 10,
    borderRadius: 50,
    alignItems: "center",
    justifyContent: "center",
    alignSelf: "center",
  },
  button: {
    backgroundColor: "#000",
    paddingHorizontal: 20,
    paddingVertical: 10,
    marginTop: 10,
    borderRadius: 50,
  },
  buttonText: {
    color: "#fff",
    fontSize: 18,
  },
});

export const darkStyles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#0a0a0a",
    alignItems: "center",
    justifyContent: "space-between",
    paddingVertical: 80,
  },
  title: {
    fontFamily: "UnifrakturMaguntia",
    fontSize: 40,
    textAlign: "center",
    color: "#fff",
  },
  text: {
    fontSize: 18,
    color: "#fff",
  },
  shareBtn: {
    padding: 10,
    borderRadius: 50,
    alignItems: "center",
    justifyContent: "center",
    alignSelf: "center",
  },
  button: {
    backgroundColor: "#fff",
    paddingHorizontal: 20,
    paddingVertical: 10,
    marginTop: 10,
    borderRadius: 50,
  },
  buttonText: {
    color: "#000",
    fontSize: 18,
  },
});
