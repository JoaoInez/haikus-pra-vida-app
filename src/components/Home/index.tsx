import React from "react";
import {
  Text,
  View,
  TouchableOpacity,
  Share,
  ActivityIndicator,
} from "react-native";
import { StatusBar } from "expo-status-bar";
import { useFonts } from "expo-font";
import { useQuery, gql } from "@apollo/client";
import { useColorScheme } from "react-native-appearance";
import { Ionicons } from "@expo/vector-icons";

import { haiku } from "types/haiku";
import { defaultStyles, lightStyles, darkStyles } from "./styles";

export const HAIKU_QUERY = gql`
  query haiku {
    haiku {
      topVerse
      middleVerse
      bottomVerse
    }
  }
`;

const Home = () => {
  const [loaded] = useFonts({
    UnifrakturMaguntia: require("../../assets/fonts/UnifrakturMaguntia-Regular.ttf"),
  });

  const colorScheme = useColorScheme();

  const { loading, data, refetch } = useQuery<haiku>(HAIKU_QUERY, {
    fetchPolicy: "cache-and-network",
  });

  const onGenHaiku = () => refetch();

  const onShare = async () => {
    try {
      await Share.share({
        message: `${data?.haiku?.topVerse}\n${data?.haiku?.middleVerse}\n${data?.haiku?.bottomVerse}`,
      });
    } catch (error) {}
  };

  const getThemeStyles = (): typeof lightStyles | typeof darkStyles =>
    colorScheme === "light" || colorScheme === "no-preference"
      ? lightStyles
      : darkStyles;

  if (!loaded) return null;

  return (
    <View style={getThemeStyles().container}>
      <StatusBar
        style={
          colorScheme === "light" || colorScheme === "no-preference"
            ? "dark"
            : "light"
        }
      />
      {loading && (
        <View style={defaultStyles.loadingContainer}>
          <ActivityIndicator
            color={
              colorScheme === "light" || colorScheme === "no-preference"
                ? "black"
                : "white"
            }
          />
        </View>
      )}
      <Text style={getThemeStyles().title}>Haikus p'ra Vida</Text>
      <View style={defaultStyles.textContainer}>
        {data?.haiku && (
          <>
            <Text style={getThemeStyles().text}>{data.haiku.topVerse}</Text>
            <Text style={getThemeStyles().text}>{data.haiku.middleVerse}</Text>
            <Text style={getThemeStyles().text}>{data.haiku.bottomVerse}</Text>
            <View style={defaultStyles.shareContainer}>
              <TouchableOpacity
                style={getThemeStyles().shareBtn}
                onPress={onShare}
              >
                <Ionicons
                  name="md-share"
                  size={24}
                  color={
                    colorScheme === "light" || colorScheme === "no-preference"
                      ? "black"
                      : "white"
                  }
                />
              </TouchableOpacity>
            </View>
          </>
        )}
      </View>
      <TouchableOpacity
        style={getThemeStyles().button}
        onPress={onGenHaiku}
        disabled={loading}
      >
        <Text style={getThemeStyles().buttonText}>Demonimiza</Text>
      </TouchableOpacity>
    </View>
  );
};

export default Home;
