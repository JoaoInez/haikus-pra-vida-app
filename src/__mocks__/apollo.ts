import { HAIKU_QUERY } from "../components/Home";

const defaultMocks = [
  {
    request: {
      query: HAIKU_QUERY,
    },
    result: {
      data: {
        haiku: {
          topVerse: "Testa-me a sanidade",
          middleVerse: "Este lixo falha sempre",
          bottomVerse: "Quero desistir",
        },
      },
    },
  },
];

export { defaultMocks };
