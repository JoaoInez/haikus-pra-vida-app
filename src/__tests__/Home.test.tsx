import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import { MockedProvider } from "@apollo/client/testing";
import { AppearanceProvider } from "react-native-appearance";

import Home from "../components/Home";
import { defaultMocks } from "../__mocks__/apollo";

describe("<Home />", () => {
  it("matches snapshots", async () => {
    const tree = render(
      <MockedProvider mocks={defaultMocks} addTypename={false}>
        <AppearanceProvider>
          <Home />
        </AppearanceProvider>
      </MockedProvider>
    );

    expect(tree).toMatchSnapshot();

    await waitFor(() => {});

    expect(tree).toMatchSnapshot();
  });

  it("fetches haikus", async () => {
    const { getByText } = render(
      <MockedProvider mocks={defaultMocks} addTypename={false}>
        <AppearanceProvider>
          <Home />
        </AppearanceProvider>
      </MockedProvider>
    );

    await waitFor(() => {});

    expect(getByText("Testa-me a sanidade")).toBeTruthy();
    expect(getByText("Este lixo falha sempre")).toBeTruthy();
    expect(getByText("Quero desistir")).toBeTruthy();
  });
});
