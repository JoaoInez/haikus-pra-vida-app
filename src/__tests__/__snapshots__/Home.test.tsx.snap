// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> matches snapshots 1`] = `
<RNCAppearanceProvider
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`<Home /> matches snapshots 2`] = `
<RNCAppearanceProvider
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0a0a0a",
        "flex": 1,
        "justifyContent": "space-between",
        "paddingVertical": 80,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "textAlign": "center",
          }
        }
      >
        Haikus p'ra Vida
      </Text>
      <Text
        style={
          Object {
            "color": "#FF006E",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "left": 2,
            "position": "absolute",
            "textAlign": "center",
            "top": 2,
            "zIndex": -1,
          }
        }
      >
        Haikus p'ra Vida
      </Text>
      <Text
        style={
          Object {
            "color": "#2A9D8F",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "left": 4,
            "position": "absolute",
            "textAlign": "center",
            "top": 4,
            "zIndex": -2,
          }
        }
      >
        Haikus p'ra Vida
      </Text>
      <Text
        style={
          Object {
            "color": "#FFBE0B",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "left": 6,
            "position": "absolute",
            "textAlign": "center",
            "top": 6,
            "zIndex": -3,
          }
        }
      >
        Haikus p'ra Vida
      </Text>
      <Text
        style={
          Object {
            "color": "#3A86FF",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "left": 8,
            "position": "absolute",
            "textAlign": "center",
            "top": 8,
            "zIndex": -4,
          }
        }
      >
        Haikus p'ra Vida
      </Text>
      <Text
        style={
          Object {
            "color": "#8338EC",
            "fontFamily": "UnifrakturMaguntia",
            "fontSize": 40,
            "left": 10,
            "position": "absolute",
            "textAlign": "center",
            "top": 10,
            "zIndex": -5,
          }
        }
      >
        Haikus p'ra Vida
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "position": "relative",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
          }
        }
      >
        Testa-me a sanidade
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
          }
        }
      >
        Este lixo falha sempre
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
          }
        }
      >
        Quero desistir
      </Text>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "marginTop": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "borderRadius": 50,
              "justifyContent": "center",
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 24,
                },
                undefined,
                Object {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Ôçπ
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 50,
          "marginTop": 10,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 18,
          }
        }
      >
        Demonimiza
      </Text>
    </View>
  </View>
</RNCAppearanceProvider>
`;
